mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/memory/DeleteByQueryExecutor.java

```diff
@@ -0,0 +1,88 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.memory;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.task.Task;
+import org.apache.james.util.FunctionalUtils;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class DeleteByQueryExecutor {
+    static class Notifiers {
+        private final Runnable userHandledNotifier;
+        private final Runnable searchErrorNotifier;
+        private final Runnable deletionErrorNotifier;
+        private final Runnable permanentlyDeletedMessageNotifyer;
+
+        Notifiers(Runnable userHandledNotifier, Runnable searchErrorNotifier, Runnable deletionErrorNotifier, Runnable permanentlyDeletedMessageNotifyer) {
+            this.userHandledNotifier = userHandledNotifier;
+            this.searchErrorNotifier = searchErrorNotifier;
+            this.deletionErrorNotifier = deletionErrorNotifier;
+            this.permanentlyDeletedMessageNotifyer = permanentlyDeletedMessageNotifyer;
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeleteByQueryExecutor.class);
+
+    private final DeletedMessageVault deletedMessageVault;
+
+    public DeleteByQueryExecutor(DeletedMessageVault deletedMessageVault) {
+        this.deletedMessageVault = deletedMessageVault;
+    }
+
+    public Task.Result deleteByQuery(Query allBeforeRetentionPeriod, Notifiers notifiers) {
+        return Flux.from(deletedMessageVault.usersWithVault())
+            .flatMap(user -> deleteByQueryForUser(allBeforeRetentionPeriod, user, notifiers))
+            .reduce(Task::combine)
+            .blockOptional()
+            .orElse(Task.Result.COMPLETED);
+    }
+
+    private Mono<Task.Result> deleteByQueryForUser(Query allBeforeRetentionPeriod, User user, Notifiers notifiers) {
+        return Flux.from(deletedMessageVault.search(user, allBeforeRetentionPeriod))
+            .flatMap(message -> deleteMessage(user, message.getMessageId(), notifiers))
+            .onErrorResume(e -> {
+                LOGGER.error("Error encountered while searching old mails in {} vault", user.asString(), e);
+                notifiers.searchErrorNotifier.run();
+                return Mono.just(Task.Result.PARTIAL);
+            })
+            .reduce(Task::combine)
+            .map(FunctionalUtils.toFunction(() -> LOGGER.info("Retention applied for {} vault", user.asString())))
```

Comment:
why can't you use `do*` functions?

Reply:
I thought they were asynchronous and that you didn't liked it!